### PR TITLE
Enable sub quantum delay

### PR DESCRIPTION
**Pull request by b-ma**
Hey, here is a first draft on #79

Don't merge this yet, I'm really not sure it's finished but It would be nice to have your feedback on what I have done so far. As this is related to the graph, I'm a bit nervous because I'm not sure I fully understand the `visit/order` stuff :)

- I had to remove a test which is failing now but that should be reviewed because it was checking against clamped delay values. I'm not sure why it breaks, need to dig more on that.
- Also I'd like to add more tests to be sure things behaves as expected

Except from the one I mentioned the all other tests seems to be happy

### CODE REVIEW

File: src/render/graph.rs [hunk 1]
```diff
@@ -180,13 +183,32 @@ impl Graph {
...
+                let node = self.nodes.get(node_id).unwrap();
+                node.processor.can_break_cycle()
+            });
+
+            match cycle_breaker_node {
+                Some(&node_id) => {
+                    // store node id to clear the node outgoing edges
+                    cycle_breakers.push(node_id);
+                    // remove nodes from mark temp after pos
+                    marked_temp.truncate(pos);
```
Comment (b-ma) [inline src/render/graph.rs:201]: I'm not completely sure of this

Comment (orottier) [reply src/render/graph.rs:201]: Yeah, it's quite hard to reason about it. My gut feeling is that this implementation is to optimistic. It would be safer to clear out `marked_temp` and also remove these nodes from `marked`. Then start visiting again from the first item that was in `marked_temp`

**Review by orottier — COMMENTED** [review]
Great to see we are nearing a solution for this issue.
Very happy to see the extensive tests for it. It will help us iterate on the solution while checking all works well.
I dropped two notes, but I do not have a solution ready at hand. Would you like me to take the work from here and spend some time on it, or will you have another go?

File: src/render/processor.rs [hunk 2]
```diff
@@ -26,6 +26,10 @@ pub struct RenderScope {
 ///
 /// Check the `examples/worklet.rs` file for example usage of this trait.
 pub trait AudioProcessor: Send {
+    fn can_break_cycle(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (orottier) [reply src/render/processor.rs:29]: We cannot really use downcasting for this problem, e.g. https://stackoverflow.com/a/33687996 without adding an `as_any` method to all processors, which is undesirable too.

Another solution would be to explicitly store `is_cycle_breaker` in the `Node`, and setting that value with a call  
`ConcreteBaseAudioContext::markCycleBreaker(&self, node_id: &AudioNodeId)` that passes the info to the render thread

[the remark this answers is not in the document]
Comment (b-ma) [reply src/render/processor.rs:29]: Yup I agree that's not very clean, especially as it only concerns one (weird) node, that was just the only I found :)

File: src/render/graph.rs (hunk 1 above)
Comment (b-ma) [reply src/render/graph.rs:201]: Maybe a nice thing to have would be a unit test with really weird edge case: some cycle inside a cycle I guess, to check this behaves as expected (whatever the solution)?

**Comment by b-ma** [comment]
> Would you like me to take the work from here and spend some time on it, or will you have another go?

Actually if you have ideas to iterate, please go on yes! I can try to contribute the test I spoke about but then I'm short with ideas about how to continue

**Comment by orottier** [comment]
Okay I will have a try at all three things (`can_break_cycle`, tests, and correct sorting algorithm). I have some ideas

**Comment by b-ma** [comment]
> Okay I will have a try at all three things (can_break_cycle, tests, and correct sorting algorithm). I have some ideas

ok cool!

**Comment by b-ma** [comment]
[7ef0597](https://github.com/orottier/web-audio-api-rs/pull/198/commits/7ef05970c8ad9c2181c365b07b6216be9689ba26) Nice approach, way cleaner!

**Comment by b-ma** [comment]
The graph test suite looks very nice!

**Comment by orottier** [comment]
Okay this is conceptually finished but I'm still looking for a more elegant graph ordering after the cycles are broken

**Comment by b-ma** [comment]
I think if we have a test like:

```
            +---------+    +---------+   
            v         |    v         |
+---+     +---+     +----------+   +---+
| 1 | --> | 2 | --> | 3: delay |   | 4 |
+---+     +---+     +----------+   +---+
            |            |           |
            v            +------------
          +---+
          | 5 |
          +---+
```

where `4` should be muted while `3` should break the cycle and that it works consistently over several iterations. My impression is that we have nice test case to know if the optimistic solution is enough (i.e. truncate only is ok, which I actually I can't think why it would break but maybe I'm wrong) or if we need to be pessimistic (reorder the graph, considering that creating feedback delays is not something people do so often that is it a big issue neither :)

To make it even more perverse and see how the recursive stuff behaves, we could also add connections between `4` and `2`:

```
           +--------------------------+
           |                          |
           | +--------+    +--------+ |  
           v v        |    v        | |
+---+     +---+     +----------+   +---+
| 1 | --> | 2 | --> | 3: delay |   | 4 |
+---+     +---+     +----------+   +---+
            |            |           |
            v            +------------
          +---+
          | 5 |
          +---+
```
or reverse

```
            +---------+     +--------+   
            v         |     v        | 
+---+     +---+     +----------+   +---+
| 1 | --> | 2 | --> | 3: delay |   | 4 |
+---+     +---+     +----------+   +---+
           | |        |             | |
           | |        +-------------+ |
           | +------------------------+           
           v            
          +---+
          | 5 |
          +---+
```

 which I guess should lead to the exact same result (`4` muted, `3` breaks cycle)?

**Comment by orottier** [comment]
I'm adding your examples, stay tuned!
Unfortunately I already concluded the optimistic graph sorting is incorrect at [b1498a2](https://github.com/orottier/web-audio-api-rs/pull/198/commits/b1498a24d5da84a78e9143ee64d5595f186ed93a)
This still has the `truncate and continue` style, but the cycle integration test consistently fails
I tried to fix the optimistic method but could not get it to work yet, I think you need to unroll the marked, marked_temp and ordered items in a very specific way. Maybe I will succeed tomorror

**Comment by orottier** [comment]
L2 access seems to be improved with [b85de76](https://github.com/orottier/web-audio-api-rs/pull/198/commits/b85de766535dfeb607b694e3a97c736973fc4a9d)
In that vein I opened #208 to explore further optimizations

File: src/node/delay.rs [hunk 3]
```diff
@@ -432,7 +435,9 @@ impl AudioProcessor for DelayReader {
         let quantum_duration = RENDER_QUANTUM_SIZE as f64 * dt;
 
         let delay_param = params.get(&self.delay_time);
-        let in_cycle = self.in_cycle.load(Ordering::SeqCst);
+
+        let latest_frame_written = self.latest_frame_written.load(Ordering::SeqCst);
+        let in_cycle = latest_frame_written != scope.current_frame;
```
Comment (b-ma) [inline src/node/delay.rs:440]: I'm not that sure this is right when in cycle actually:
- if not in cycle the order of rendering is guaranteed, therefore we can trust this check
- but when in cycle, the order of rendering is not guaranteed (at least it wasn't, did it changed somehow?), so the check could fail sometime?

**Comment by b-ma** [comment]
> This still has the truncate and continue style, but the cycle integration test consistently fails

oh ok, I didn't see the sub comment in the commit message, so pessimistic we must be indeed!

> L2 access seems to be improved with [b85de76](https://github.com/orottier/web-audio-api-rs/pull/198/commits/b85de766535dfeb607b694e3a97c736973fc4a9d)
> In that vein I opened https://github.com/orottier/web-audio-api-rs/issues/208 to explore further optimizations

Nice one! (perf is really a trap :)

File: src/node/delay.rs (hunk 3 above)
Comment (orottier) [reply src/node/delay.rs:440]: Hmmmm right, it is quite nuanced actually and my code needs to be improved
- When not in a cycle, there is a edge from writer to reader so writer always renders first
- When in a cycle, that edge is dropped. The nature of the cycle means output of the reader feeds into the writer, hence, the reader always renders first
- However, when the user breaks the cycle manually by dropping some other connections, the order is no longer guaranteed and thus random

The last case is a bit of an edge case of course, but I could make the in_cycle property sticky (one you are in a cycle, treat as if you will always stay in that cycle) so we prevent erratic behaviour in rendering of sub-quantum delays.

Another issue I realize is when the Writer is dropped, it will stop updating the `latest_frame_written`. This causes the Reader to set `in_cycle` true, disallowing sub-quantum delays. If the delay is set to 100 samples this behaviour is not right becase the final 28 frames om the last Writer call should still render directly in the first 28 frames of the next Reader output.

Pfft, not simple stuff, I will iterate again

File: src/node/constant_source.rs [hunk 4]
```diff
@@ -186,7 +186,8 @@ impl AudioProcessor for ConstantSourceRenderer {
             current_time += dt;
         }
 
-        true
+        // tail_time false when output has ended this quantum
+        stop_time > next_block_time
```
Comment (b-ma) [inline src/node/constant_source.rs:190]: Remembered that we had to output one channel of silence before returning false, but can't find it in the spec actually... I think the check line 165 makes no sense anymore then?

Comment (b-ma) [reply src/node/constant_source.rs]: handled in e088c95

**Comment by orottier** [comment]
Thanks for the update, I think we have done enought in this PR.
I added two more test cases, one remark:

> Something like that should result in both 4 muted and 3 breaks cycle: (D)

Because 2 is also part of the cycle with 4, it is muted as well
